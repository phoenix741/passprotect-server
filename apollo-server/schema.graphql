scalar Date

# Encrypted content
type EncryptedContent {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

# Encrypted content
input EncryptedContentInput {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

type Error {
  fieldName: String,
  message: String
}

type Errors {
  errors: [Error]!
}

# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

union RegisterUserResult = User | Errors

# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

union CreateUpdateLineResult = WalletLine | Errors

# Information needed to connect to the server and retreive object from database.
input ConnectionInformationInput {
  # Login of the user
  username: String!,
  # Password used for connection
  password: String!
}

# Result of the connection
type ConnectionInformation {
  # Token to be used in connection information
  token: String,
  # The connected user
  user: User!
}

union CreateSessionResult = ConnectionInformation | Errors

# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

type RootQuery {
  # Request a user
  user(id: ID!): User,

    # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]

  # Request the current session
  session: User

  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult

  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors

  createSession(input: ConnectionInformationInput!): CreateSessionResult
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}

schema {
  query: RootQuery
  mutation: RootMutation
  subscription: RootSubscription
}
