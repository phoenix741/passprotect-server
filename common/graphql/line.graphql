# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

extend type RootQuery {
  # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]
}

union CreateUpdateLineResult = WalletLine | Errors

extend type RootMutation {
  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors
}
