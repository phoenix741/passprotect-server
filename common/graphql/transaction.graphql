# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

extend type RootQuery {
  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}
