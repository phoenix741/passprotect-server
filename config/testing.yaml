config:
    mongodb:
        host: 'mongodb://localhost:27017/passprotect'
        user: ~
        password: ~
        options: ~
