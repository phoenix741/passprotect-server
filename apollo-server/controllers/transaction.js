import { checkPermission } from '../utils/authentification'
import { getTransactions } from '../services/transaction'
import { pubsub, TRANSACTION_ADDED_TOPIC } from '../services/subscriptions'
import { withFilter } from 'graphql-subscriptions'
import moment from 'moment'

export const resolvers = {
  RootQuery: {
    async transactions (obj, { earliest }, { user }) {
      checkPermission(user)
      earliest = moment(earliest).toDate()

      return getTransactions(user, { earliest })
    }
  },

  RootSubscription: {
    transactionAdded: {
      subscribe: withFilter(() => pubsub.asyncIterator(TRANSACTION_ADDED_TOPIC), (payload, context) => {
        return payload.transactionAdded.user === context.user._id
      })
    }
  }
}
