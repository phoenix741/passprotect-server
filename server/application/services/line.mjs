'use strict';

import debug from 'debug';
import {NotFoundError} from '../models/exception';
import {getLines as getLinesModel, getLine as getLineModel, saveLine as saveLineModel, removeLine as removeLineModel} from '../models/line';
import {createTransaction} from '../services/transaction';

const log = debug('App:Service:Line');

export function getLines(user) {
	log('Get all lines from user ', user._id);

	const filter = {};
	filter.user = user._id;

	const sort = {};
	sort.group = 1;
	sort.label = 1;

	return getLinesModel(filter, sort);
}

export function getLine(id) {
	log(`Get the line with id ${id}`);
	return getLineModel(id);
}

export function saveLine(line) {
	log(`Create the line of type ${line.type}`);

	line.updatedAt = new Date();

	return getLineIfAvailable(line._id, line._rev)
		.then(function (oldLine) {
			return saveLineModel(line).tap(newLine => createTransaction('line', oldLine, newLine));
		});
}

export function removeLine(id) {
	log(`Remove the line with the id ${id}`);

	return getLineIfAvailable(id)
		.tap(() => removeLineModel(id))
		.tap(oldLine => createTransaction('line', oldLine));
}

function getLineIfAvailable(id, rev) {
	return getLineModel(id, rev).catch(NotFoundError, () => null);
}
