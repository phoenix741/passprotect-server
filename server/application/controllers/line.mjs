import {omit, pick, includes, isString, isEmpty} from 'lodash'
import expressPromiseRouter from 'express-promise-router'
import i18n from 'i18next'
import fs from 'fs'
import path from 'path'
import debug from 'debug'
import {authenticate, permission, checkPermission} from '../utils/passport'
import {getLines, getLine, removeLine, saveLine} from '../services/line'
import {getGroups} from '../services/group'

const log = debug('App:Controllers:Line')

const router = expressPromiseRouter()
export default [authenticate(), permission(), router]

log('Load line type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', '..', 'common', 'graphql', 'line.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    lines (obj, args, {user}) {
      return checkPermission(user).then(() => getLines(user)).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    },

    line (obj, {id}, {user}) {
      return getLine(id).tap(line => checkPermission(user, [], line.user)).then(filterLine)
    },

    groups (obj, args, {user}) {
      return checkPermission(user).then(() => getGroups(user))
    }
  },
  User: {
    lines (obj, args, {user}) {
      return getLines(obj._id).map(line => checkPermission(user, [], line.user).return(line)).map(filterLine)
    }
  },

  RootMutation: {
    createUpdateLine (obj, {input}, {user}) {
      let checkPermissionPromise = Promise.resolve(input)
      if (input._id) {
        checkPermissionPromise = getLine(input._id)
          .tap(line => checkPermission(user, [], line.user))
          .then(line => {
            input._id = line._id
            return input
          })
      }

      input.user = user._id
      return checkPermissionPromise
        .then(input => sanitizeInput(input))
        .then(data => saveLine(data))
        .then(filterLine)
        .catch(parseErrors)
    },

    removeLine (obj, {id}, {user}) {
      return getLine(id)
        .tap(line => checkPermission(user, [], line.user))
        .then(() => removeLine(id))
        .then(() => ({errors: []}))
        .catch(parseErrors)
    }
  },

  CreateUpdateLineResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }
      return 'WalletLine'
    }
  }
}

/**
 * @apiDefine text Informations to encrypt (client side) for text
 */
/**
 * @apiDefine password Informations to encrypt (client side) for password
 */
/**
 * @apiDefine card Informations to encrypt (client side) for payment card
 */

/**
 * @apiDefine UpdateLine
 *
 * @apiParam {String="text","card","password"} type type of line
 * @apiParam {String} label label of the line
 * @apiParam {Object} encryption Informations used for encrypted data
 * @apiParam {Object} encryption.informations Encrypted informations
 * @apiParam {String} encryption.informations.authTag Authentification tag of the encrypted information
 * @apiParam {String} encryption.informations.content Encrypted informations
 * @apiParam {String} encryption.salt Salt used to encrypt/decrypt informations
 *
 * @apiParam (text) {String} text Text that will be encrypted
 * @apiParam (text) {String} notes Custom note added to the line
 *
 * @apiParam (password) {String} username Username associated to the pasword
 * @apiParam (password) {String} password The password
 * @apiParam (password) {String} siteUrl The site where the password is used
 * @apiParam (password) {String} notes Custom note added to the line
 *
 * @apiParam (card) {String="VISA","Carte bleu","Mastercard","American Express","Maestro"} type Type of card
 * @apiParam (card) {String} nameOnCard Name of the person on the card
 * @apiParam (card) {String} cardNumber The card number
 * @apiParam (card) {String} cvv The code in the back on the card
 * @apiParam (card) {String} expiry The expiration date
 * @apiParam (card) {String} code The code to use to pay with the card
 * @apiParam (card) {String} notes Custom note added to the line
 *
 * @apiParamExample {json} Line:
 *   {
 *     "type": "text",
 *     "label": "My custom text",
 *     "encryption": {
 * )      "salt": "83b01169a90fe764",
 *       "informations": {
 *         "content": "d13a32829c2547b8bca6da5bb568d3",
 *         "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *       }
 *     }
 *   }
 */

/**
 * @apiDefine ResponseLineComplete
 *
 * @apiSuccess {String} _id Id of the line
 * @apiSuccess {String} type type of line
 * @apiSuccess {String} label label of the line
 * @apiSuccess {Object} encryption Informations used for encrypted data
 * @apiSuccess {Object} encryption.informations Encrypted informations
 * @apiSuccess {String} encryption.informations.content Encrypted content of the line
 * @apiSuccess {String} encryption.informations.authTag Authentification tag of the encrypted informations
 * @apiSuccess {String} encryption.salt Salt used to encrypt informations
 *
 *
 * @apiSuccessExample {json} Success text:
 *   HTTP/1.1 200 OK
 *     {
 *       "_id": "57685936f4495bc62c4d28a3",
 *       "type": "text",
 *       "label": "My custom text",
 *       "encryption": {
 *         "salt": "83b01169a90fe764",
 *         "informations": {
 *           "content": "d13a32829c2547b8bca6da5bb568d3",
 *           "authTag": "db47515bdcb7643484d60189b7ad4e33"
 *         }
 *       }
 *     }
 */

router.route('')
  /**
   * @api {get} /api/lines Request all lines
   *
   * @apiDescription Request the list of lines from the connected users. Only the _id, type, and label of the line is
   * given. No encrypted information are send in this method.
   *
   * @apiName GetLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiParam {Number} [offset=0] Offset for pagination
   * @apiParam {Number} [limit=10000] Limit of items used for pagination
   *
   * @apiSuccess {String} _id Id of the line
   * @apiSuccess {String} type type of line
   * @apiSuccess {String} label label of the line
   *
   * @apiSuccessExample Success:
   *     HTTP/1.1 200 OK
   *     [
   *       {
   *         "_id": "57685936f4495bc62c4d28a3",
   *         "type": "card",
   *         "label": "My credit card"
   *       },
   *       {
   *         "_id": "57685936f4495bc62c4d28a4",
   *         "type": "password",
   *         "label": "My password"
   *       },
   *     ]
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   * @apiError Forbidden Your are not allowed to access to this API
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   *
   * @apiErrorExample {json} Forbidden:
   *     HTTP/1.1 403 Forbidden
   */
  .get((req, res) => {
    return getLines(req.user).map(filterLine).then(res.json.bind(res))
  })

  /**
   * @api {post} /api/lines Create new line
   * @apiName PostLines
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   *
   * @apiError Unauthorized Only the connected user can access to list of lines
   *
   * @apiErrorExample {json} Unauthorized:
   *     HTTP/1.1 401 Unauthorized
   */
  .post((req, res) => {
    const data = omit(req.body, '_id', 'user')

    return checkAndSaveLine(data, req, res)
  })

router.param('lineId', (req, res, next, lineId) => {
  getLine(lineId, req.user)
    .then(line => {
      // Can't access or update lines that isn't associated to the account of the user
      if (line.user !== req.user._id) {
        const notAuthorizedError = new Error('Not authorized to access to thhis subscriptions')
        notAuthorizedError.status = 403
        return Promise.reject(notAuthorizedError)
      }

      req.line = line
      return next()
    })
    .catch(next)
})

/**
 * @api {get} /api/lines/:lineId Request line
 * @apiName GetLine
 * @apiGroup Line
 * @apiPermission ROLE_USER
 *
 * @apiParam {String} lineId The id of the line
 *
 * @apiUse ResponseLineComplete
 *
 * @apiError Unauthorized Only the connected user can access to list of lines
 * @apiError Forbidden Your are not allowed to access to this API
 *
 * @apiErrorExample {json} Unauthorized:
 *     HTTP/1.1 401 Unauthorized
 *
 * @apiErrorExample {json} Forbidden:
 *     HTTP/1.1 403 Forbidden
 */
router.route('/:lineId')
  .get((req, res) => res.json(filterLine(req.line)))
  /**
   * @api {put} /api/lines/:lineId Update the line
   * @apiName PutLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiUse UpdateLine
   * @apiUse ResponseLineComplete
   */
  .put((req, res) => {
    // Force the id of the line to ensure that the user doesn't try to change it
    const data = omit(req.body, '_id')
    data._id = req.line._id

    return checkAndSaveLine(data, req, res)
  })
  /**
   * @api {delete} /api/lines/:lineId Delete the line
   * @apiName DeleteLine
   * @apiGroup Line
   * @apiPermission ROLE_USER
   *
   * @apiSuccessExample {json} Success-Response:
   *     HTTP/1.1 204 OK
   */
  .delete((req, res) => removeLine(req.line._id).then(() => res.status(204).send()))

/**
 * Filter information that the user is authorized to view in the response
 * @param {Object} line The response
 * @returns {Object} The filtered object.
 */
function filterLine (line) {
  return pick(line, '_id', 'user', 'type', 'label', 'group', 'encryption', 'updatedAt', '_rev')
}

function sanitizeInput (input) {
  const data = pick(input, '_id', 'user', 'type', 'label', 'group', 'encryption', '_rev')

  const validationError = new Error()
  validationError.status = 400

  const typeChoices = ['card', 'password', 'text']
  if (!includes(typeChoices, data.type)) {
    validationError.message = i18n.t('error:line.400.type', {choices: typeChoices.join(', ')})
    return Promise.reject(validationError)
  }

  if (!isString(data.label) || isEmpty(data.label)) {
    validationError.message = i18n.t('error:line.400.label')
    return Promise.reject(validationError)
  }

  return Promise.resolve(data)
}

/**
 * Create or modify a line given by the request.
 * @param {Object} data Object to create or modify
 * @param {Object} req Request use to read the user
 * @param {Object} res Response used to send the response.
 * @returns {Promise} Promise that the line will be saved.
 */
function checkAndSaveLine (data, req, res) {
  // Force the user stored in the line.
  data.user = req.user._id

  return sanitizeInput(data)
    .then(data => saveLine(data))
    .then(line => res.json(filterLine(line)))
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'global',
      message: err.message
    }]
  }
}
