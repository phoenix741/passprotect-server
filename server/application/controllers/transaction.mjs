import {checkPermission} from '../utils/passport'
import {getTransactions} from '../services/transaction'
import moment from 'moment'
import fs from 'fs'
import path from 'path'
import debug from 'debug'

const log = debug('App:Controllers:Session')

log('Load transaction type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', '..', 'common', 'graphql', 'transaction.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    async transactions (obj, {earliest}, {user}) {
      checkPermission(user)
      earliest = moment(earliest).toDate()

      return getTransactions(user, {earliest})
    }
  },

  RootSubscription: {
    transactionAdded (transaction) {
      return transaction
    }
  }
}

export const setupFunctions = {
  transactionAdded ({context}) {
    return {
      transactionAdded: {
        filter: transaction => transaction.user === context.user._id
      }
    }
  }
}
