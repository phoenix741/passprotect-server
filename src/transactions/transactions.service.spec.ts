import { Test } from '@nestjs/testing';
import { getModelToken } from '@nestjs/mongoose';
import * as mongoose from 'mongoose';
import mockingoose from 'mockingoose';
import { TransactionSchema } from './schemas/transaction.schema';
import {
  TransactionsService,
  TRANSACTION_ADDED_TOPIC,
} from './transactions.service';
import { ObjectID } from 'bson';
import { TransactionTypeEnum } from '../shared/interfaces/transaction-type-enum.interface';
import { LineEntity } from '../lines/models/line.entity';
import { LineTypeEnum } from '../shared/interfaces/line-type-enum.interface';
import { TransactionEntity } from './models/transaction.entity';

describe('TransactionsService', () => {
  let transactionModel;
  let transactionsService;

  const line: LineEntity = {
    _id: new ObjectID('5ce15929f02fc23595ec38f1'),
    type: LineTypeEnum.card,
    label: 'Mon label',
    group: 'Mon group',
    encryption: {
      authTag: Buffer.from('authTag'),
      content: Buffer.from('content'),
      iv: Buffer.from('iv'),
      salt: Buffer.from('salt'),
    },
    user: 'userId',
    createdAt: new Date(1558224000),
    updatedAt: new Date(1558224000),
  } as LineEntity;

  const transaction: TransactionEntity = {
    _id: new ObjectID('5ce15929f02fc23595ec38f0'),
    type: TransactionTypeEnum.line,
    line: new ObjectID('5ce15929f02fc23595ec38f1'),
    user: 'userId',
    before: null,
    after: line,
    sha512: 'sha512',
    createdAt: new Date(1558224000),
    updatedAt: new Date(1558224000),
  } as TransactionEntity;

  beforeEach(async () => {
    transactionModel = mongoose.model('Transaction', TransactionSchema);

    const module = await Test.createTestingModule({
      providers: [
        TransactionsService,
        {
          provide: getModelToken('Transaction'),
          useValue: transactionModel,
        },
      ],
    }).compile();

    transactionsService = module.get<TransactionsService>(TransactionsService);
  });

  describe('findAll', () => {
    const doc = {
      _id: new ObjectID('5ce15929f02fc23595ec38f0'),
      type: TransactionTypeEnum.line,
      line: new ObjectID('5ce15929f02fc23595ec38f1'),
      user: 'userId',
      sha512: 'sha512',
      createdAt: new Date(1558224000),
      updatedAt: new Date(1558224000),
    };

    it('without date', async () => {
      const userId = 'userId';
      const finderMock = query => {
        expect(query.getQuery()).toMatchSnapshot('find query without date');
        return [doc];
      };
      mockingoose(transactionModel).toReturn(finderMock, 'find');
      const lines = await transactionsService.findAll(userId);
      expect(lines.map(line => line.toObject())).toMatchSnapshot(
        'result of find all without date',
      );
    });

    it('with date', async () => {
      const userId = 'userId';
      const finderMock = query => {
        expect(query.getQuery()).toMatchSnapshot('find query without date');
        return [doc];
      };
      mockingoose(transactionModel).toReturn(finderMock, 'find');
      const lines = await transactionsService.findAll(userId, {
        earliest: new Date(1558224000),
      });
      expect(lines.map(line => line.toObject())).toMatchSnapshot(
        'result of find all with date',
      );
    });
  });

  describe('createTransaction', () => {
    it('success', async () => {
      const createMock = transactionToCreate => {
        expect(transactionToCreate.toObject()).toMatchSnapshot(
          {
            _id: expect.any(ObjectID),
            createdAt: expect.any(Date),
            updatedAt: expect.any(Date),
          },
          'transaction to create',
        );
        return transaction;
      };
      mockingoose(transactionModel).toReturn(createMock, 'save');
      const createdTransaction = await transactionsService.createTransaction(
        TransactionTypeEnum.line,
        null,
        line,
      );
      expect(createdTransaction.toObject()).toMatchSnapshot(
        'transaction created',
      );
    });

    it('failed', async () => {
      mockingoose(transactionModel).toReturn(new Error('failed'), 'save');
      await expect(
        transactionsService.createTransaction(
          TransactionTypeEnum.line,
          null,
          line,
        ),
      ).rejects.toThrow('failed');
    });

    it('failed - no line', async () => {
      await expect(
        transactionsService.createTransaction(
          TransactionTypeEnum.line,
          null,
          null,
        ),
      ).rejects.toThrow('Transaction: before or after should be defined');
    });
  });

  describe('transactionPubSub', () => {
    it('publish', async done => {
      mockingoose(transactionModel).toReturn(transaction, 'save');
      const unsubscribeId = await transactionsService.transactionPubSub.subscribe(
        TRANSACTION_ADDED_TOPIC,
        t => {
          expect(t).toMatchSnapshot('result of published transaction');
          transactionsService.transactionPubSub.unsubscribe(unsubscribeId);
          done();
        },
      );
      transactionsService.createTransaction(
        TransactionTypeEnum.line,
        null,
        line,
      );
    });
  });
});
