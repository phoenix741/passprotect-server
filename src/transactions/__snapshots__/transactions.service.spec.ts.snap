// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsService createTransaction success: transaction created 1`] = `
Object {
  "_id": "5ce15929f02fc23595ec38f0",
  "after": Object {
    "_id": "5ce15929f02fc23595ec38f1",
    "createdAt": 1970-01-19T00:50:24.000Z,
    "encryption": Object {
      "authTag": Object {
        "data": Array [
          97,
          117,
          116,
          104,
          84,
          97,
          103,
        ],
        "type": "Buffer",
      },
      "content": Object {
        "data": Array [
          99,
          111,
          110,
          116,
          101,
          110,
          116,
        ],
        "type": "Buffer",
      },
      "iv": Object {
        "data": Array [
          105,
          118,
        ],
        "type": "Buffer",
      },
      "salt": Object {
        "data": Array [
          115,
          97,
          108,
          116,
        ],
        "type": "Buffer",
      },
    },
    "group": "Mon group",
    "label": "Mon label",
    "type": "card",
    "updatedAt": 1970-01-19T00:50:24.000Z,
    "user": "userId",
  },
  "before": null,
  "createdAt": 1970-01-19T00:50:24.000Z,
  "line": "5ce15929f02fc23595ec38f1",
  "sha512": "sha512",
  "type": "line",
  "updatedAt": 1970-01-19T00:50:24.000Z,
  "user": "userId",
}
`;

exports[`TransactionsService createTransaction success: transaction to create 1`] = `
Object {
  "_id": Any<ObjectID>,
  "after": Object {
    "_id": "5ce15929f02fc23595ec38f1",
    "createdAt": 1970-01-19T00:50:24.000Z,
    "encryption": Object {
      "authTag": Object {
        "data": Array [
          97,
          117,
          116,
          104,
          84,
          97,
          103,
        ],
        "type": "Buffer",
      },
      "content": Object {
        "data": Array [
          99,
          111,
          110,
          116,
          101,
          110,
          116,
        ],
        "type": "Buffer",
      },
      "iv": Object {
        "data": Array [
          105,
          118,
        ],
        "type": "Buffer",
      },
      "salt": Object {
        "data": Array [
          115,
          97,
          108,
          116,
        ],
        "type": "Buffer",
      },
    },
    "group": "Mon group",
    "label": "Mon label",
    "type": "card",
    "updatedAt": 1970-01-19T00:50:24.000Z,
    "user": "userId",
  },
  "before": null,
  "createdAt": Any<Date>,
  "line": "5ce15929f02fc23595ec38f1",
  "sha512": "b2d1d285b5199c85f988d03649c37e44fd3dde01e5d69c50fef90651962f48110e9340b60d49a479c4c0b53f5f07d690686dd87d2481937a512e8b85ee7c617f",
  "type": "line",
  "updatedAt": Any<Date>,
  "user": "userId",
}
`;

exports[`TransactionsService findAll with date: find query without date 1`] = `
Object {
  "updatedAt": Object {
    "$gte": 1970-01-19T00:50:24.000Z,
  },
  "user": "userId",
}
`;

exports[`TransactionsService findAll with date: result of find all with date 1`] = `
Array [
  Object {
    "_id": "5ce15929f02fc23595ec38f0",
    "createdAt": 1970-01-19T00:50:24.000Z,
    "line": "5ce15929f02fc23595ec38f1",
    "sha512": "sha512",
    "type": "line",
    "updatedAt": 1970-01-19T00:50:24.000Z,
    "user": "userId",
  },
]
`;

exports[`TransactionsService findAll without date: find query without date 1`] = `
Object {
  "user": "userId",
}
`;

exports[`TransactionsService findAll without date: result of find all without date 1`] = `
Array [
  Object {
    "_id": "5ce15929f02fc23595ec38f0",
    "createdAt": 1970-01-19T00:50:24.000Z,
    "line": "5ce15929f02fc23595ec38f1",
    "sha512": "sha512",
    "type": "line",
    "updatedAt": 1970-01-19T00:50:24.000Z,
    "user": "userId",
  },
]
`;

exports[`TransactionsService transactionPubSub publish: result of published transaction 1`] = `
Object {
  "transactionAdded": Object {
    "_id": "5ce15929f02fc23595ec38f0",
    "after": Object {
      "_id": "5ce15929f02fc23595ec38f1",
      "createdAt": 1970-01-19T00:50:24.000Z,
      "encryption": Object {
        "authTag": Object {
          "data": Array [
            97,
            117,
            116,
            104,
            84,
            97,
            103,
          ],
          "type": "Buffer",
        },
        "content": Object {
          "data": Array [
            99,
            111,
            110,
            116,
            101,
            110,
            116,
          ],
          "type": "Buffer",
        },
        "iv": Object {
          "data": Array [
            105,
            118,
          ],
          "type": "Buffer",
        },
        "salt": Object {
          "data": Array [
            115,
            97,
            108,
            116,
          ],
          "type": "Buffer",
        },
      },
      "group": "Mon group",
      "label": "Mon label",
      "type": "card",
      "updatedAt": 1970-01-19T00:50:24.000Z,
      "user": "userId",
    },
    "before": null,
    "createdAt": 1970-01-19T00:50:24.000Z,
    "line": "5ce15929f02fc23595ec38f1",
    "sha512": "sha512",
    "type": "line",
    "updatedAt": 1970-01-19T00:50:24.000Z,
    "user": "userId",
  },
}
`;
